Error when requesting keyword_view: RESOURCE_EXHAUSTED: gRPC message exceeds maximum size 4194304

Hi,

when doing this request
`SELECT customer.id, ad_group.id, ad_group_criterion.criterion_id, ad_group_criterion.url_custom_parameters, ad_group_criterion.keyword.text, ad_group_criterion.keyword.match_type, ad_group_criterion.final_urls, segments.date, metrics.clicks, metrics.cost_micros FROM keyword_view WHERE metrics.clicks > 0 AND segments.date BETWEEN '2022-02-13' AND '2022-02-14'`
for one specific account I get this exception 
```
Caused by: com.google.api.gax.rpc.ResourceExhaustedException: io.grpc.StatusRuntimeException: RESOURCE_EXHAUSTED: gRPC message exceeds maximum size 4194304: 4769258
	at com.google.api.gax.rpc.ApiExceptionFactory.createException(ApiExceptionFactory.java:57)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:72)
	at com.google.api.gax.grpc.GrpcApiExceptionFactory.create(GrpcApiExceptionFactory.java:60)
	at com.google.api.gax.grpc.ExceptionResponseObserver.onErrorImpl(ExceptionResponseObserver.java:82)
	at com.google.api.gax.rpc.StateCheckingResponseObserver.onError(StateCheckingResponseObserver.java:86)
	at com.google.api.gax.grpc.GrpcDirectStreamController$ResponseObserverAdapter.onClose(GrpcDirectStreamController.java:149)
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39)
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23)
	at io.grpc.ForwardingClientCallListener$SimpleForwardingClientCallListener.onClose(ForwardingClientCallListener.java:40)
	at com.google.ads.googleads.lib.logging.LoggingInterceptor$1$1.onClose(LoggingInterceptor.java:113)
	at io.grpc.internal.DelayedClientCall$DelayedListener$3.run(DelayedClientCall.java:463)
	at io.grpc.internal.DelayedClientCall$DelayedListener.delayOrExecute(DelayedClientCall.java:427)
	at io.grpc.internal.DelayedClientCall$DelayedListener.onClose(DelayedClientCall.java:460)
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:557)
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:69)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:738)
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:717)
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:133)
```

I think this is an extreme edge case - I do this request for some hundred accounts and it happens only at one single account. I haven't seen this error before, today 13:12 CET was the first time. The error occurs when using API v9 and v10 (I use google-ads 17.0.0). When I change the query to 
`SELECT ad_group_criterion.resource_name, ad_group_criterion.url_custom_parameters, ad_group_criterion.keyword.text, ad_group_criterion.keyword.match_type, ad_group_criterion.final_urls, segments.date, metrics.clicks, metrics.cost_micros FROM keyword_view WHERE metrics.clicks > 0 AND segments.date BETWEEN '2022-02-13' AND '2022-02-14'`
I get no error and the maximum of GoogleAdsRow.toByteArray().length is 497 bytes.

Unfortunately the logs don't contain a requestId, the RequestLogger only logs
`15:20:45.915 [WARN]  [grpc-default-executor-0] [c.g.a.g.l.l.RequestLogger:159] : FAILURE REQUEST SUMMARY. Method: google.ads.googleads.v10.services.GoogleAdsService/SearchStream, Endpoint: googleads.googleapis.com:443, CustomerID: MY_CUSTOMER_ID, RequestID: null, ResponseCode: RESOURCE_EXHAUSTED, Fault: gRPC message exceeds maximum size 4194304: 4769258.`

Best regards,

Christian

--- a/google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java
+++ b/google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java
@@ -70,6 +70,18 @@ public abstract class GoogleAdsClient extends AbstractGoogleAdsClient {
   /** The default endpoint for Google Ads API services. */
   private static final String DEFAULT_ENDPOINT = "googleads.googleapis.com:443";
 
+  /**
+   * Default max inbound metadata (headers) size. Inbound headers for the Google Ads API may exceed
+   * the default max of 8KB.
+   */
+  @VisibleForTesting static final Integer DEFAULT_INBOUND_METADATA_SIZE = 16 * 1024 * 1024;
+
+  /**
+   * Default max inbound message (response) size. Large response will often exceed the default of
+   * 4MB.
+   */
+  @VisibleForTesting static final Integer DEFAULT_INBOUND_MESSAGE_SIZE = 64 * 1024 * 1024;
+
   static {
     // Alpha feature to optimize the client startup time.
     Primer.primeBasicsIfEnabled();
@@ -88,12 +100,8 @@ public abstract class GoogleAdsClient extends AbstractGoogleAdsClient {
                             new RequestLogger(),
                             clientBuilder.getHeaders(),
                             clientBuilder.getEndpoint())))
-            // Issue 131: inbound headers may exceed default (8kb) max header size.
-            // Sets max header size to 16MB, which should be more than necessary.
-            .setMaxInboundMetadataSize(16 * 1024 * 1024)
-            // Sets max response size to 64MB, since large responses will often exceed the default
-            // (4MB).
-            .setMaxInboundMessageSize(64 * 1024 * 1024)
+            .setMaxInboundMetadataSize(DEFAULT_INBOUND_METADATA_SIZE)
+            .setMaxInboundMessageSize(DEFAULT_INBOUND_MESSAGE_SIZE)
             .build();
     clientBuilder
         .setEndpoint(DEFAULT_ENDPOINT)
@@ -584,6 +592,8 @@ public abstract class GoogleAdsClient extends AbstractGoogleAdsClient {
                         ImmutableList.of(
                             new LoggingInterceptor(
                                 new RequestLogger(), getHeaders(), getEndpoint())))
+                .setMaxInboundMetadataSize(DEFAULT_INBOUND_METADATA_SIZE)
+                .setMaxInboundMessageSize(DEFAULT_INBOUND_MESSAGE_SIZE)
                 .build();
       }
       if (transportChannelProvider.needsHeaders()) {

--- a/google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java
+++ b/google-ads/src/main/java/com/google/ads/googleads/lib/GoogleAdsClient.java
@@ -74,13 +74,13 @@ public abstract class GoogleAdsClient extends AbstractGoogleAdsClient {
    * Default max inbound metadata (headers) size. Inbound headers for the Google Ads API may exceed
    * the default max of 8KB.
    */
-  @VisibleForTesting static final Integer DEFAULT_INBOUND_METADATA_SIZE = 16 * 1024 * 1024;
+  @VisibleForTesting static final Integer DEFAULT_MAX_INBOUND_METADATA_SIZE = 16 * 1024 * 1024;
 
   /**
    * Default max inbound message (response) size. Large response will often exceed the default of
    * 4MB.
    */
-  @VisibleForTesting static final Integer DEFAULT_INBOUND_MESSAGE_SIZE = 64 * 1024 * 1024;
+  @VisibleForTesting static final Integer DEFAULT_MAX_INBOUND_MESSAGE_SIZE = 64 * 1024 * 1024;
 
   static {
     // Alpha feature to optimize the client startup time.
@@ -100,8 +100,8 @@ public abstract class GoogleAdsClient extends AbstractGoogleAdsClient {
                             new RequestLogger(),
                             clientBuilder.getHeaders(),
                             clientBuilder.getEndpoint())))
-            .setMaxInboundMetadataSize(DEFAULT_INBOUND_METADATA_SIZE)
-            .setMaxInboundMessageSize(DEFAULT_INBOUND_MESSAGE_SIZE)
+            .setMaxInboundMetadataSize(DEFAULT_MAX_INBOUND_METADATA_SIZE)
+            .setMaxInboundMessageSize(DEFAULT_MAX_INBOUND_MESSAGE_SIZE)
             .build();
     clientBuilder
         .setEndpoint(DEFAULT_ENDPOINT)
@@ -592,8 +592,8 @@ public abstract class GoogleAdsClient extends AbstractGoogleAdsClient {
                         ImmutableList.of(
                             new LoggingInterceptor(
                                 new RequestLogger(), getHeaders(), getEndpoint())))
-                .setMaxInboundMetadataSize(DEFAULT_INBOUND_METADATA_SIZE)
-                .setMaxInboundMessageSize(DEFAULT_INBOUND_MESSAGE_SIZE)
+                .setMaxInboundMetadataSize(DEFAULT_MAX_INBOUND_METADATA_SIZE)
+                .setMaxInboundMessageSize(DEFAULT_MAX_INBOUND_MESSAGE_SIZE)
                 .build();
       }
       if (transportChannelProvider.needsHeaders()) {

--- a/google-ads/src/test/java/com/google/ads/googleads/lib/GoogleAdsClientTest.java
+++ b/google-ads/src/test/java/com/google/ads/googleads/lib/GoogleAdsClientTest.java
@@ -39,6 +39,7 @@ import com.google.ads.googleads.v10.services.SearchGoogleAdsStreamRequest;
 import com.google.ads.googleads.v10.services.SearchGoogleAdsStreamResponse;
 import com.google.api.gax.grpc.GaxGrpcProperties;
 import com.google.api.gax.grpc.GrpcStatusCode;
+import com.google.api.gax.grpc.InstantiatingGrpcChannelProvider;
 import com.google.api.gax.grpc.testing.LocalChannelProvider;
 import com.google.api.gax.grpc.testing.MockServiceHelper;
 import com.google.api.gax.rpc.ApiClientHeaderProvider;
@@ -190,10 +191,7 @@ public class GoogleAdsClientTest {
 
     // Build a new client from the file.
     GoogleAdsClient client =
-        GoogleAdsClient.newBuilder()
-            .fromPropertiesFile(propertiesFile)
-            .setTransportChannelProvider(localChannelProvider)
-            .build();
+        GoogleAdsClient.newBuilder().fromPropertiesFile(propertiesFile).build();
     assertGoogleAdsClient(client, null, true);
   }
 
@@ -253,7 +251,7 @@ public class GoogleAdsClientTest {
 
   /** Tests building a client without the use of a properties file. */
   @Test
-  public void buildWithoutPropertiesFile_supportsAllFields() throws IOException {
+  public void buildWithoutPropertiesFile_supportsAllFields() {
     Credentials credentials =
         UserCredentials.newBuilder()
             .setClientId(CLIENT_ID)
@@ -265,7 +263,6 @@ public class GoogleAdsClientTest {
             .setCredentials(credentials)
             .setDeveloperToken(DEVELOPER_TOKEN)
             .setLoginCustomerId(LOGIN_CUSTOMER_ID)
-            .setTransportChannelProvider(localChannelProvider)
             .build();
     assertGoogleAdsClient(client, true);
   }
@@ -886,6 +883,20 @@ public class GoogleAdsClientTest {
           "Scope",
           Collections.singleton("https://www.googleapis.com/auth/adwords"),
           serviceAccountCredentials.getScopes());
+    }
+
+    if (client.getTransportChannelProvider() == client.getDefaultTransportChannelProvider()) {
+      InstantiatingGrpcChannelProvider channelProvider =
+          (InstantiatingGrpcChannelProvider) client.getTransportChannelProvider();
+      assertEquals(
+          "Max inbound metadata size",
+          GoogleAdsClient.DEFAULT_INBOUND_METADATA_SIZE,
+          channelProvider.getMaxInboundMetadataSize());
+      assertEquals(
+          "Max inbound message size",
+          GoogleAdsClient.DEFAULT_INBOUND_MESSAGE_SIZE,
+          // For some reason, this setting is only available on the builder.
+          channelProvider.toBuilder().getMaxInboundMessageSize());
     }
 
     assertEquals("Developer token", DEVELOPER_TOKEN, client.getDeveloperToken());

--- a/google-ads/src/test/java/com/google/ads/googleads/lib/GoogleAdsClientTest.java
+++ b/google-ads/src/test/java/com/google/ads/googleads/lib/GoogleAdsClientTest.java
@@ -890,11 +890,11 @@ public class GoogleAdsClientTest {
           (InstantiatingGrpcChannelProvider) client.getTransportChannelProvider();
       assertEquals(
           "Max inbound metadata size",
-          GoogleAdsClient.DEFAULT_INBOUND_METADATA_SIZE,
+          GoogleAdsClient.DEFAULT_MAX_INBOUND_METADATA_SIZE,
           channelProvider.getMaxInboundMetadataSize());
       assertEquals(
           "Max inbound message size",
-          GoogleAdsClient.DEFAULT_INBOUND_MESSAGE_SIZE,
+          GoogleAdsClient.DEFAULT_MAX_INBOUND_MESSAGE_SIZE,
           // For some reason, this setting is only available on the builder.
           channelProvider.toBuilder().getMaxInboundMessageSize());
     }

